include/tensor.cuh

```diff
@@ -16,45 +15,43 @@
 /**
  * Define defaults
  */
-
-#define TENSOR_DEFAULT_TYPE double
+#define DEFAULT_FPX double
 #define THREADS_PER_BLOCK 512
 #define DIM2BLOCKS(n) ((n) / THREADS_PER_BLOCK + ((n) % THREADS_PER_BLOCK != 0))
 #if (__cplusplus >= 201703L)  ///< if c++17 or above
-#define TENSOR_TEMPLATE_WITH_TYPE template<typename T = TENSOR_DEFAULT_TYPE>
+#define TEMPLATE_WITH_TYPE_T template<typename T = DEFAULT_FPX>
 #else
-#define TENSOR_TEMPLATE_WITH_TYPE template<typename T>
+#define TEMPLATE_WITH_TYPE_T template<typename T>
 #endif
 #if (__cplusplus >= 202002L)  ///< if c++20 or above
-#define TENSOR_REQUIRES_TYPE requires std::floating_point<T>
+#define TEMPLATE_CONSTRAINT_REQUIRES_FPX requires std::floating_point<T>
 #else
-#define TENSOR_REQUIRES_TYPE
+#define TEMPLATE_CONSTRAINT_REQUIRES_FPX
 #endif
 
 /**
  * Check for errors when calling GPU functions
  */
-#define gpuErrChk(status) { gpuAssert((status), std::source_location::current()); }
+#define gpuErrChk(status) { gpuAssert((status), __FILE__, __LINE__); } while(false)
```

Comment:
while(false) ?

Reply:
@ruairimoran  This strange `while(false)` covers a weird corner case; without it you can't do this:

```c++
if (condition)
   gpuErrChk(blah); gpuErrChk(blah); 
```

i.e., combine two `gpuErrChk` in one line under an `if` or `for` without curly brackets.